File: web/src/types/track.ts
import type { Tag } from "./tag";

export type TrackStatus = "uploaded" | "processing" | "ready" | "failed" | string;

export type Track = {
  id: number;
  title: string;
  artist: string | null;
  album: string | null;
  duration_seconds: number | null;
  content_type: string;
  original_file_path: string | null;
  playback_file_path: string | null;
  cover_path: string | null;
  source_url: string | null;
  format: string | null;
  bitrate: number | null;
  status: TrackStatus;
  processing_job_status: string | null;
  processing_error_message: string | null;
  liked: boolean;
  cooldown_until: string | null;
  created_at: string;
  updated_at: string;
  tags: Tag[];
};

export type TrackUpdate = {
  title?: string | null;
  artist?: string | null;
  album?: string | null;
  content_type?: string | null;
  source_url?: string | null;
  liked?: boolean | null;
  cooldown_until?: string | null;
  tag_ids?: number[] | null;
};

export type TrackMetadataUpdate = Pick<
  TrackUpdate,
  | "album"
  | "artist"
  | "content_type"
  | "cooldown_until"
  | "liked"
  | "source_url"
  | "title"
>;

export type TrackTagUpdate = Pick<TrackUpdate, "tag_ids">;

File: backend/app/schemas/track.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict

from app.schemas.tag import TagResponse


class TrackUpdate(BaseModel):
    title: str | None = None
    artist: str | None = None
    album: str | None = None
    content_type: str | None = None
    source_url: str | None = None
    liked: bool | None = None
    cooldown_until: datetime | None = None
    tag_ids: list[int] | None = None


class TrackResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    title: str
    artist: str | None
    album: str | None
    duration_seconds: int | None
    content_type: str
    original_file_path: str | None
    original_file_size_bytes: int | None
    original_file_sha256: str | None
    playback_file_path: str | None
    playback_file_sha256: str | None
    cover_path: str | None
    source_url: str | None
    format: str | None
    bitrate: int | None
    normalized_metadata_key: str | None
    status: str
    processing_job_status: str | None = None
    processing_error_message: str | None = None
    liked: bool
    cooldown_until: datetime | None
    created_at: datetime
    updated_at: datetime
    tags: list[TagResponse]
